### DISCUSSION TITLE
Airflow variables through helm chart

### DISCUSSION THREAD

**Discussion by fernhtls** [apache/airflow#16387] [Q&A] [answered]
Hi all,

We are currently evaluating migrating our current Airflow helm deployment using the stable/airflow-helm community chart (https://github.com/airflow-helm/charts/tree/main/charts/airflow) to the apache-airflow helm chart (https://github.com/apache/airflow/tree/main/chart), as it includes a lot of improvements, like **pgbouncer** and so on, and it's fully prepared for Airflow 2.*.

Only thing it seems to be missing for now is how to inject variables.

With the stable/airflow-helm chart we can inject in the deployment our Airflow variables:

https://github.com/helm/charts/blob/f4f301ae450101b981805bd045451f08c0d74afa/stable/airflow/values.yaml#L285

```
scheduler:
  variables: |
    {
      "myvar1": "value1",
      "myvar2": "value2",
      ....
    }
```

In the new version of the airflow-helm community chart, it seems to be still there, but not below the `scheduler` values anymore, but below `airflow`:
https://github.com/airflow-helm/charts/blob/9961aa0921c9ab1ec606208c17ff26ec1b27e832/charts/airflow/values.yaml#L126

**I could not find anything equivalent in the values.yaml or anything in the rest of the documentation for the apache-airflow/airflow chart.**

**Could anyone point to me if this is supported or not?**

Cheers,

Fernando

**Comment by fernhtls** [comment, marked as answer]
Just replying to my own question, so in this case I will go with my own initContainer, getting a json template and just importing it into airflow, but I think it would be good to include it directly in the chart.
Even tough values that could be passed through the template, as `{{ template "airflow_image" . }}` and others, I will need to put them manually in the `extraInitContainers`.

**Reply by fernhtls** [reply]
On checking further we can't push any `extraInitContainers`, and we can't change the `scheduler` `args` to include for example a `airflow variables import <json.file>`, or even have a job. Would there be any other suggestions to still do it though this helm chart? PS: The one from the community creates actually creates a job to import the variables.

**Reply by potiuk** [reply]
I do not think it is a good idea. This is again the same case as in other discussions - installing packages, adding variables adding connections etc. at container start should NOT be part of the chart. @kaxil and @jedcunningham  to confirm.

Generally speaking, adding variables should be done outside of installation of chart - similarly as creating users, and maintaining DAGs. Helm chart is really to update application versions, and variable configuration is not part of it. You can think of helm as "apt-get" interface to update your application to a newer version.  Would you like to get newer version of a dependency installed which would override your configured parameters with "original" values? Probably not a good idea. 

Same here. There are a number of scenarios where setting or updating variables in the chart might lead to problems. The main problem is that If you update your variables in the chart - you probably expect the variable to be updated, but if you manually modified your variables, this means that by updating Airflow to newer version, you would override your manually set variables. Bad idea.

If you need to configure your variables via import, you should keep it completely separately from the chart (as a "variable configuration service", not "application installation service").

**Reply by fernhtls** [reply]
Hi @potiuk,

Thanks for replying.

I agree completely that it's not something really part of the installation of airflow, but more like a `runtime` configuration.

In our case we only change things through code, so our infra, airflow and so on is being pushed through **terraform**, including the variables.

It's very convenient to have it versioned and pushed through code in our case, of course nothing prevents someone from changing it manually through the UI as well (in our case most users are only Viewers, but I'm not 100% sure if this is blocked by the **rbac** roles, or if it could be).

In our case I was looking for something similar to the community chart, so a k8s job that gets triggered to create / update those variables, and in the chart itself they have some flags to control to not update the vars, as explained [here](https://github.com/airflow-helm/charts/blob/9961aa0921c9ab1ec606208c17ff26ec1b27e832/charts/airflow/values.yaml#L126).

And I think the job `create-variables` itself for the community chart can be seem as some sort of **"variable configuration service"**, and it would be using the same image, configmaps, secrets and so on already created / maintained by the chart itself.

[Airflow helm community chart here job](https://github.com/airflow-helm/charts/blob/main/charts/airflow/templates/jobs/job-create-variables.yaml).

Thanks anyway, and I might just continue to use the community helm chart anyway :) .

Cheers,

Fernando

**Reply by potiuk** [reply]
> It's very convenient to have it versioned and pushed through code in our case, of course nothing prevents someone from changing it manually through the UI as well (in our case most users are only Viewers, but I'm not 100% sure if this is blocked by the **rbac** roles, or if it could be).
> In our case we only change things through code, so our infra, airflow and so on is being pushed through terraform, including the variables.

Absolutely no denying here. And thanks for sharing your use case. I think the discussion here is very valuable to us to understand how people are using the chart. I am sure @kaxil and @jedcunningham will agree with me that while we have our opinions, we listen to the users and we are addressing their needs if we see that they are valid (but maybe in a bit different way than the non-official chart. So I hope we can learn from the discussion and propose something that will be good according to the principles we started with and good for the users.

I just question the approach where this is done by an init job in helm chart. Just to repeat it - I believe we have a few of abstraction layers, and each layer has different responsibility. Making the Helm chart responsible to manage all the layers is simply trying to follow a "god" pattern in software architecture where everything is managed and controlled from a single location - sort of "monolithic approach" for the application. However K8S & friends philosophy is a bit different and much more "layered" and shows it true potential when - instead of centralizing your application deployment in one place it benefits from distributing it across the layers - because then you can manage each of the layers separately in the pace, life-cycle and approach that is best (and different) for each layer. 

I think this is a bit of fallacy of the non-official chart that it tries to make the helm chart the "God" managing the deployment, where a better (in my opinion) approach is to distribute the deployment to different layers. While it seems convenient first to manage everything via the chart, it's not a very good approach from the maintainability point of view because lifecycles of different layers are very different.

The layers I see now in Airflow (from slowest to fastest rate of changes/lifecycle). Also any change to "higher" layer should not require modification of the "lower" layer. The other way it might happen - if you modify "lower" layer, often it implies some changes or at lest "restart/redeploy" at the "higher" layers:

1) Metadata Database (Postgres/MySQL instance) -> should be maintained separately, and outside of K8S (K8S is not good for persistence). Lifecycle here is super-slow. You set it up once and that's pretty much unless you want to change clustering, DB configuration etc. Not related to Airflow versions usually. 

2) Airflow Container Image -> should be maintained separately by building your own image. This changes more frequently - when you upgrade Airflow version but also when you add new deployment-wide dependencies. Depending on your Pull Policy (Always), you do not even have to re-install Helm Chart if you update your image Though it's likely smarter to build your image with specific version and update the Helm chart with the new version when you want to update it to get better control over it. The images should be immutable when it comes to dependencies.

3) Airflow Application: Airflow Code + Pods + Matadata Schema  + environment  configuration - > this is all part of the "Airflow application" as I see it. This is the only layer that should be managed by the Helm Chart. You should reinstall Helm chart whenever the image changes (barring the Always Pull Policy) and whenever you make any changes in the config environment, or whenever new version of Airflow is released (which also implies image rebuild so new version of Airflow goes one layer down). But there should be no need of reinstalling application when you change variables, connections, users, RBAC configuration, DAGs etc. 

4) Connection and Variables configuration in meta-data. Those change much more frequently - for example every time password is rotated or configuration of particular part of your "business model" changes. This is not part of the "Airflow application". This is part of your "Business configuration" and your "External services" that you connect to. And maintaining those should not be related to maintaining "application". 

5) User/RBAC management - this is something that should also be done outside of Airflow Application and it is separate layer. It can use airflow image to run airflow commands but as such it does not have to imply helm reinstalling - it can be done completely separately. It changes every time the users change. It can be separated out via LDAP/Oauth etc. integration and fully integrated in your existing authentication mechanisms.

6) Dags - those change very rapidly, possibly hourly. Solutions like Git-Sync is the right way to approach distribution of DAGs, so that there is no need usually to restart or re-deploy images, charts, or configuration.

And here Terraform might be the "glue" for you to combine management of layers 1 - 5 (and even setup the git repositories and access for layer 6) in the right way including layer dependencies. Also in the way that will make updates to only the layers that are affected. For example you should easily make your Terraform dependencies work in the way that whenever you change your image definition, it will automatically build the image, replace configuration of the image in your helm chart and reinstalls it.  

There are a number of options you can use for your layers 4) 5):  you can for example manage a separate K8S job for this in Terraform: https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/job and probably there are already other plugins that can do the job nicely and Terraform can add nice dependencies there. I simply think using Helm Chart for that is a bad idea. You already have Terraform to manage your infrastructure as a code. Why do you want to make Helm Chart do the job that it was never created for ?

**Reply by fernhtls** [reply]
Hi @potiuk ,

Totally agree with the separation of layers.

If you observe the community chart, [it's actually a k8s Job](https://github.com/airflow-helm/charts/blob/main/charts/airflow/templates/jobs/job-create-users.yaml) as I stated before.

Yes, sure i can glue everything on my side as well, but then I'll have to attach secrets for the metadata db, and other configs as well to achieve it, and it would be out of the templating of the chart, so a bit of manual work, that in case when I need to change the image / version or whatever I would need to port it as well to this configuration management.

I know I have a big number of options here, and my idea of the **initContainers** are not the best probably, but the job for creating those variables is a total separated layer, to apply the configuration.

Cheers,

Fernando

**Reply by potiuk** [reply]
> 
> If you observe the community chart, [it's actually a k8s Job](https://github.com/airflow-helm/charts/blob/main/charts/airflow/templates/jobs/job-create-users.yaml) as I stated before.
> 
> Yes, sure i can glue everything on my side as well, but then I'll have to attach secrets for the metadata db, and other configs as well to achieve it, and it would be out of the templating of the chart, so a bit of manual work, that in case when I need to change the image / version or whatever I would need to port it as well to this configuration management.

I wonder what others think here, but my main point is that managing configuration/variables/users simply follows completely different lifecycle - you don't necessarily want to re-run variable creation when you update Airflow version/chart. On the other hand you probably do not want to install new chart version when you just want to update variables. The approach in the non-official chart make those two lifecycles closely-coupled and the flags whether to run the update or not are super-confusing. They are transient (per operation) flags rather than stateful flags and you have to remember to set them properly and they do not follow the terraform's 'apply' semantics, where you only describe the "target" state. 

If you are using Terraform, then most likely you already have your TF variables defined for secrets, connection strings, etc. and you do not really need all the extra parameters that helm chart needs to run such import even if you use "airflow import" statement.  for that you can easily just need image and connection string to database and it will be much more convenient to run it standalone, because then you can mount your .json import file separately only for that job and you do not really need to even think how to do in the helm-chart - it is not needed anywhere else in the chart.

You do not really need advanced templating. The [kustomize](https://kubernetes.io/blog/2018/05/29/introducing-kustomize-template-free-configuration-customization-for-kubernetes/) which is pretty much integral part of k8s should be more than enough to run such a job via terraform using those variables and then you can control the life-cycle of configuration independently from controlling the application lifecycle (which is what you really want). And you can even do it in the way that you can detect and react somehow to potential conflicts when such configuration is updated manually/using different channel.

I am afraid the approach the non-official helm chart introduced prefers "convenience" of keeping configuration in a single place vs the "maintainability" and "life-cycle management correctness". As @kaxil already mentioned before in similar discussions  - "convenient" does not always mean "correct".  Narrowing the set of tools you use to helm chart only is not the right choice in the k8s world I am afraid.

**Reply by jedcunningham** [reply]
I personally am not a fan of the `import` approach, most succinctly because it opens the door to the "which is the right value?" problem when someone modifies it in the db. As @potiuk pointed out, doing a `helm upgrade` would reset the value.

However, other options exist to have your variables set via helm. The easiest is using env vars via `env` in your values. Even if someone creates a variable of the same name in the db, the env var wins.

Another option is using the file based secrets backend via `extraVolumes` and `extraVolumeMounts` for the various components.

I'll keep it short, but check out those options. I'm happy to expand on them if you'd like me to.

**Reply by potiuk** [reply]
Good points @jedcunningham. I think the point with secret/env based variables is very good, because then it allows you to define a truly "managed" values - i.e. you do not have to solve the problem of conflicts. There is one source of truth only.

**Reply by fernhtls** [reply]
hi @jedcunningham ,

Thanks for your point on this.

Unfortunately for our case env vars would not be a good option, as we have more then 50 variables configured.

Our problem is not control the single font of truth for the vars, as it will always be the json we define in terraform, because just a few people here would be Admins, and would have access to set / change the vars. The ETL / job developers, they are only Viewers, so if they want to change what's deployed, it needs to pass through terraform.

I've got already to create a k8s job, in the same fashion as the community helm chart, outside of the apache-airflow helm chart, using the definitions I got from the webserver / scheduler containers, and it's working fine. If we change the json the job is destroyed / recreated, and it runs again updating the vars.

The only thing is if those secret names, key names, if they change we will have to update the job, but a minor work really.

My intention in asking here mainly was to see if in some way, this would be integrated in the helm chart, for not having to update the job in the future manually, but it might that only us are asking / doing things like this.

For now the job outside the helm chat works as a charm.

Thanks,

Fernando

**Reply by kaxil** [reply]
Yeah, I agree with @potiuk and @jedcunningham , you can use Environment Variables using `env`/`secret` -- so no exposing your Airflow Variables with secrets as plain-text.


> On checking further we can't push any `extraInitContainers`, and we can't change the `scheduler` `args` to include for example a `airflow variables import <json.file>`, or even have a job. Would there be any other suggestions to still do it though this helm chart? PS: The one from the community creates actually creates a job to import the variables.

You will now be able to override the scheduler command from next version of Airflow Helm Chart (to be released within a month):

https://github.com/apache/airflow/blob/3f674bd6bdb281cd4c911f8b1bc7ec489a24c49d/chart/values.yaml#L465-L467

However, I would definitely suggest you to try the environment variable route.

**Reply by kaxil** [reply]
The current chart and future improvement will make our chart very extensible (similar to Airflow) itself where the defaults work with best practises and have things required to run a min deployment but would allow extending the Chart for users who know what they are doing :)

**Reply by fernhtls** [reply]
hi @kaxil , 

It's great to hear that it will have more flexibility, I've tried as well to change the args but the schema didn't allow as well.

I do agree that in some cases the secret / envFrom approach would be much better, but now we have Dags already developed and with dependency on those variables.

If the numbers were different, less variables and less dags depending on it, I would change it to secret / envFrom, or event just mount it.

Thanks you all for your view and support on this!

Fernando

**Reply by kaxil** [reply]
Hi @fernhtls Just to be also clear that you can also define a "Airflow Variable" using "Environment Variable" as documented in https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html#storing-variables-in-environment-variables,

example:

```shell
export AIRFLOW_VAR_FOO=BAR

# To use JSON, store them as JSON strings
export AIRFLOW_VAR_FOO_BAZ='{"hello":"world"}'
```

and you can use them in your DAGs as:

```python
from airflow.models import Variable
foo = Variable.get("foo")
foo_json = Variable.get("foo_baz", deserialize_json=True)
```

This way you can use `env`/`secret` in Helm Chart and your DAGs will continue working 😄

**Reply by potiuk** [reply]
And one more thing @fernhtls - I think you might not be aware of this: https://airflow.apache.org/docs/apache-airflow/stable/security/secrets/secrets-backend/local-filesystem-secrets-backend.html 

Using Local Filesystem backend you do not even have to change anything with your .json. You just make your json available as secret and mount it in the right folder. That's it.

**Comment by fernhtls** [comment]
Hi,

I saw all this new cases of regarding the backend data and I was not really
sure about what it was. Good to know and I'll check it out for sure.

Thanks,

Fernando

**Comment by fernhtls** [comment]
Sorry **configs, not **cases. I'll check it out for sure.
